// @uniswap/universal-router/typechain/factories/LockAndMsgSender__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { LockAndMsgSender } from "../LockAndMsgSender";

export class LockAndMsgSender__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<LockAndMsgSender> {
    return super.deploy(overrides || {}) as Promise<LockAndMsgSender>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): LockAndMsgSender {
    return super.attach(address) as LockAndMsgSender;
  }
  connect(signer: Signer): LockAndMsgSender__factory {
    return super.connect(signer) as LockAndMsgSender__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): LockAndMsgSender {
    return new Contract(address, _abi, signerOrProvider) as LockAndMsgSender;
  }
}

const _abi = [
  {
    inputs: [],
    name: "ContractLocked",
    type: "error",
  },
];

const _bytecode =
  "0x60808060405234602657600080546001600160a01b031916600117905560119081602c8239f35b600080fdfe600080fdfea164736f6c6343000811000a";
