// @uniswap/universal-router/typechain/factories/TestCustomErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { TestCustomErrors } from "../TestCustomErrors";

export class TestCustomErrors__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<TestCustomErrors> {
    return super.deploy(overrides || {}) as Promise<TestCustomErrors>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): TestCustomErrors {
    return super.attach(address) as TestCustomErrors;
  }
  connect(signer: Signer): TestCustomErrors__factory {
    return super.connect(signer) as TestCustomErrors__factory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): TestCustomErrors {
    return new Contract(address, _abi, signerOrProvider) as TestCustomErrors;
  }
}

const _abi = [
  {
    inputs: [],
    name: "InvalidSignature",
    type: "error",
  },
  {
    inputs: [],
    name: "UnsafeCast",
    type: "error",
  },
];

const _bytecode =
  "0x608080604052346013576011908160198239f35b600080fdfe600080fdfea164736f6c6343000811000a";
